In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
gdp_df = pd.read_csv("../data/gdp_percapita.csv", nrows=6869)
gdp_df.head()

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN

In [2]:
internet_df = pd.read_csv('../data/internet_use.csv',nrows=4496)
internet_df.head()

Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN

7. Look at the shape of each DataFrame - how many rows? How many columns?

In [3]:
print(gdp_df.shape)

(6869, 4)


gdp_df has 6869 rows and 4 columns

In [4]:
print(internet_df.shape)

(4496, 4)


internet_df has 4496 rows and 4 columns

8. Take a look at the datatypes for the columns in each DataFrame.

In [5]:
print(gdp_df.dtypes)

Country or Area     object
Year                object
Value              float64
Value Footnotes    float64
dtype: object


In [6]:
print(internet_df.dtypes)

Country or Area     object
Year                object
Value              float64
Value Footnotes    float64
dtype: object


9. Take a look at the last 10 rows of each DataFrame in turn.

In [7]:
print(gdp_df.iloc[-10:])

     Country or Area      Year        Value  Value Footnotes
6859        Zimbabwe      1998  3653.920016              NaN
6860        Zimbabwe      1997  3580.048793              NaN
6861        Zimbabwe      1996  3520.430146              NaN
6862        Zimbabwe      1995  3226.413930              NaN
6863        Zimbabwe      1994  3263.934978              NaN
6864        Zimbabwe      1993  3033.504852              NaN
6865        Zimbabwe      1992  3054.889178              NaN
6866        Zimbabwe      1991  3426.598094              NaN
6867        Zimbabwe      1990  3324.348171              NaN
6868   footnoteSeqID  Footnote          NaN              NaN


In [8]:
print(internet_df.iloc[-10:])

     Country or Area      Year     Value  Value Footnotes
4486        Zimbabwe      2001  0.799846              NaN
4487        Zimbabwe      2000  0.401434              NaN
4488        Zimbabwe      1999  0.161676              NaN
4489        Zimbabwe      1998  0.081648              NaN
4490        Zimbabwe      1997  0.033080              NaN
4491        Zimbabwe      1996  0.016790              NaN
4492        Zimbabwe      1995  0.007684              NaN
4493        Zimbabwe      1994  0.001739              NaN
4494        Zimbabwe      1990  0.000000              NaN
4495   footnoteSeqID  Footnote       NaN              NaN


10. Drop the value footnotes column from both DataFrames. Check that this worked as expected.

In [9]:
gdp_df = gdp_df.drop(['Value Footnotes'], axis = 1)
internet_df = internet_df.drop(['Value Footnotes'], axis = 1)
print(gdp_df)
print(internet_df)

     Country or Area      Year        Value
0        Afghanistan      2019  2065.036398
1        Afghanistan      2018  2033.779002
2        Afghanistan      2017  2058.383832
3        Afghanistan      2016  2057.062164
4        Afghanistan      2015  2068.265904
...              ...       ...          ...
6864        Zimbabwe      1993  3033.504852
6865        Zimbabwe      1992  3054.889178
6866        Zimbabwe      1991  3426.598094
6867        Zimbabwe      1990  3324.348171
6868   footnoteSeqID  Footnote          NaN

[6869 rows x 3 columns]
     Country or Area      Year     Value
0        Afghanistan      2014  6.390000
1        Afghanistan      2013  5.900000
2        Afghanistan      2012  5.454545
3        Afghanistan      2011  5.000000
4        Afghanistan      2010  4.000000
...              ...       ...       ...
4491        Zimbabwe      1996  0.016790
4492        Zimbabwe      1995  0.007684
4493        Zimbabwe      1994  0.001739
4494        Zimbabwe      1990  0.000

11. Change the columns for the GDP Per Capita DataFrame to ‘Country’, ‘Year’, and ‘GDP_Per_Capita’.

In [10]:
gdp_df = gdp_df.rename(columns={"Country or Area":"Country", 'Value':'GDP_Per_Capita'})
print(gdp_df)

            Country      Year  GDP_Per_Capita
0       Afghanistan      2019     2065.036398
1       Afghanistan      2018     2033.779002
2       Afghanistan      2017     2058.383832
3       Afghanistan      2016     2057.062164
4       Afghanistan      2015     2068.265904
...             ...       ...             ...
6864       Zimbabwe      1993     3033.504852
6865       Zimbabwe      1992     3054.889178
6866       Zimbabwe      1991     3426.598094
6867       Zimbabwe      1990     3324.348171
6868  footnoteSeqID  Footnote             NaN

[6869 rows x 3 columns]


12. Change the columns for the Internet Users DataFrame to ‘Country’, ‘Year’, and ‘Internet_Users_Pct’.

In [11]:
internet_df = internet_df.rename(columns={"Country or Area": "Country", 'Value':'Internet_Users_Pct'})
print(internet_df)

            Country      Year  Internet_Users_Pct
0       Afghanistan      2014            6.390000
1       Afghanistan      2013            5.900000
2       Afghanistan      2012            5.454545
3       Afghanistan      2011            5.000000
4       Afghanistan      2010            4.000000
...             ...       ...                 ...
4491       Zimbabwe      1996            0.016790
4492       Zimbabwe      1995            0.007684
4493       Zimbabwe      1994            0.001739
4494       Zimbabwe      1990            0.000000
4495  footnoteSeqID  Footnote                 NaN

[4496 rows x 3 columns]


13. Merge the two DataFrames to one. Merge all rows from each of the two DataFrames. Call the new DataFrame gdp_and_internet_use.

In [12]:
gdp_and_internet_use = pd.merge(gdp_df, internet_df)
print(gdp_and_internet_use)

            Country      Year  GDP_Per_Capita  Internet_Users_Pct
0       Afghanistan      2014     2102.385234            6.390000
1       Afghanistan      2013     2116.465454            5.900000
2       Afghanistan      2012     2075.491747            5.454545
3       Afghanistan      2011     1904.559799            5.000000
4       Afghanistan      2010     1957.029338            4.000000
...             ...       ...             ...                 ...
3454       Zimbabwe      1996     3520.430146            0.016790
3455       Zimbabwe      1995     3226.413930            0.007684
3456       Zimbabwe      1994     3263.934978            0.001739
3457       Zimbabwe      1990     3324.348171            0.000000
3458  footnoteSeqID  Footnote             NaN                 NaN

[3459 rows x 4 columns]


14. Look at the first five rows of your new DataFrame to confirm it merged correctly.

In [13]:
print(gdp_and_internet_use.head())

       Country  Year  GDP_Per_Capita  Internet_Users_Pct
0  Afghanistan  2014     2102.385234            6.390000
1  Afghanistan  2013     2116.465454            5.900000
2  Afghanistan  2012     2075.491747            5.454545
3  Afghanistan  2011     1904.559799            5.000000
4  Afghanistan  2010     1957.029338            4.000000


15. Look at the last five rows to make sure the data is clean and as expected.

In [14]:
print(gdp_and_internet_use.tail())

            Country      Year  GDP_Per_Capita  Internet_Users_Pct
3454       Zimbabwe      1996     3520.430146            0.016790
3455       Zimbabwe      1995     3226.413930            0.007684
3456       Zimbabwe      1994     3263.934978            0.001739
3457       Zimbabwe      1990     3324.348171            0.000000
3458  footnoteSeqID  Footnote             NaN                 NaN


16. Subset the combined DataFrame to keep only the data for 2004, 2009, and 2014. Check that this happened correctly.

In [15]:
array = ['2004', '2009', '2014']
print(gdp_and_internet_use.loc[gdp_and_internet_use['Year'].isin(array)])

          Country  Year  GDP_Per_Capita  Internet_Users_Pct
0     Afghanistan  2014     2102.385234            6.390000
5     Afghanistan  2009     1758.904043            3.550000
10    Afghanistan  2004     1200.278256            0.105809
13        Albania  2014    11623.866679           60.100000
18        Albania  2009    10346.864199           41.200000
...           ...   ...             ...                 ...
3419       Zambia  2009     2917.516220            6.310000
3424       Zambia  2004     2261.766537            2.013550
3436     Zimbabwe  2014     2955.240287           19.890000
3441     Zimbabwe  2009     1925.364381           11.360000
3446     Zimbabwe  2004     2480.859342            6.564045

[482 rows x 4 columns]


17. Create three new DataFrames, one for 2004, one for 2009, and one for 2014. Give them meaningful names that aren't too long.

In [16]:
four = gdp_and_internet_use.loc[gdp_and_internet_use['Year'] == '2004']
nine = gdp_and_internet_use.loc[gdp_and_internet_use['Year'] == '2009']
fourteen = gdp_and_internet_use.loc[gdp_and_internet_use['Year'] == '2014']

print(four.head())
print(nine.head())
print(fourteen.head())

                Country  Year  GDP_Per_Capita  Internet_Users_Pct
10          Afghanistan  2004     1200.278256            0.105809
23              Albania  2004     7604.838002            2.420388
44              Algeria  2004    10058.435939            4.634475
66               Angola  2004     5593.241537            0.464815
86  Antigua and Barbuda  2004    18745.681534           24.266544
                Country  Year  GDP_Per_Capita  Internet_Users_Pct
5           Afghanistan  2009     1758.904043                3.55
18              Albania  2009    10346.864199               41.20
39              Algeria  2009    10782.361398               11.23
61               Angola  2009     7643.493749                6.00
81  Antigua and Barbuda  2009    20046.839017               42.00
                Country  Year  GDP_Per_Capita  Internet_Users_Pct
0           Afghanistan  2014     2102.385234                6.39
13              Albania  2014    11623.866679               60.10
34        

18. Which country had the highest percentage of internet users in 2014? What was the percentage? (Try typing the first 3 letters of your DataFrame name and hitting the tab key for auto-complete options).

In [17]:
fourteenhighg = fourteen.groupby(by=["Internet_Users_Pct", "Country"], dropna=False).sum()
fourteenhighgs = fourteenhighg.sort_values(["Internet_Users_Pct"], ascending=[False])
print(fourteenhighgs)

                                 GDP_Per_Capita
Internet_Users_Pct Country                     
98.16              Iceland         49652.149500
96.80              Bermuda         77361.104881
96.30              Norway          62589.373548
95.99              Denmark         52214.537753
94.67              Luxembourg     108760.916030
...                                         ...
2.10               Sierra Leone     1997.511853
1.95               Niger            1127.615541
1.72               Guinea           2061.760973
1.38               Burundi           886.235777
1.14               Timor-Leste      3264.568070

[163 rows x 1 columns]


Iceland had the highest percentage of internet users in 2014 at 98.14%.

19. Which country had the lowest percentage of internet users in 2014? What was the percentage?

In [18]:
fourteenlowg = fourteen.groupby(by=["Internet_Users_Pct", "Country"], dropna=False).sum()
fourteenlowgs = fourteenlowg.sort_values(["Internet_Users_Pct"], ascending=[True])
print(fourteenlowgs)

                                GDP_Per_Capita
Internet_Users_Pct Country                    
1.14               Timor-Leste     3264.568070
1.38               Burundi          886.235777
1.72               Guinea          2061.760973
1.95               Niger           1127.615541
2.10               Myanmar         4020.037369
...                                        ...
94.67              Luxembourg    108760.916030
95.99              Denmark        52214.537753
96.30              Norway         62589.373548
96.80              Bermuda        77361.104881
98.16              Iceland        49652.149500

[163 rows x 1 columns]


Timor-Leste had the lowest percentage of internet users in 2014 at 1.14%.

20. Repeat for 2004 and 2009.

In [19]:
fourhighg = four.groupby(by=["Internet_Users_Pct", "Country"], dropna=False).sum()
fourhighgs = fourhighg.sort_values(["Internet_Users_Pct"], ascending=[False])
print(fourhighgs)

                                GDP_Per_Capita
Internet_Users_Pct Country                    
83.890000          Sweden         44496.023401
83.880000          Iceland        45145.545658
80.930000          Denmark        50607.628971
77.690000          Norway         60893.410993
72.390000          Finland        43760.279146
...                                        ...
0.155335           Ethiopia         793.093786
0.105809           Afghanistan     1200.278256
0.077480           Tajikistan      1737.743630
0.031011           Liberia         1181.546158
0.024337           Myanmar         1696.881760

[157 rows x 1 columns]


In [20]:
fourlowg = four.groupby(by=["Internet_Users_Pct", "Country"], dropna=False).sum()
fourlowgs = fourlowg.sort_values(["Internet_Users_Pct"], ascending=[True])
print(fourlowgs)

                                GDP_Per_Capita
Internet_Users_Pct Country                    
0.024337           Myanmar         1696.881760
0.031011           Liberia         1181.546158
0.077480           Tajikistan      1737.743630
0.105809           Afghanistan     1200.278256
0.155335           Ethiopia         793.093786
...                                        ...
72.390000          Finland        43760.279146
77.690000          Norway         60893.410993
80.930000          Denmark        50607.628971
83.880000          Iceland        45145.545658
83.890000          Sweden         44496.023401

[157 rows x 1 columns]


2004: Sweden had the highest percentage of internet users at 83.89%. Myanmar had the lowest percentage of internet users at 0.02%.

In [21]:
ninehighg = nine.groupby(by=["Internet_Users_Pct", "Country"], dropna=False).sum()
ninehighgs = ninehighg.sort_values(["Internet_Users_Pct"], ascending=[False])
print(ninehighgs)

                                 GDP_Per_Capita
Internet_Users_Pct Country                     
93.000000          Iceland         48177.084123
92.080000          Norway          61886.564498
91.000000          Sweden          45638.035577
89.630000          Netherlands     51772.424928
87.310000          Luxembourg     104932.811953
...                                         ...
0.540000           Ethiopia         1150.206179
0.530000           Cambodia         2603.583778
0.260000           Sierra Leone     1372.392285
0.220000           Myanmar          2874.780227
0.185252           Timor-Leste      2724.770973

[162 rows x 1 columns]


In [22]:
ninelowg = nine.groupby(by=["Internet_Users_Pct", "Country"], dropna=False).sum()
ninelowgs = ninelowg.sort_values(["Internet_Users_Pct"], ascending=[True])
print(ninelowgs)

                                 GDP_Per_Capita
Internet_Users_Pct Country                     
0.185252           Timor-Leste      2724.770973
0.220000           Myanmar          2874.780227
0.260000           Sierra Leone     1372.392285
0.530000           Cambodia         2603.583778
0.540000           Ethiopia         1150.206179
...                                         ...
87.310000          Luxembourg     104932.811953
89.630000          Netherlands     51772.424928
91.000000          Sweden          45638.035577
92.080000          Norway          61886.564498
93.000000          Iceland         48177.084123

[162 rows x 1 columns]


2009: Iceland had the highest percentage of internet users at 93%. Timor-Leste had the lowest percentage of internet users at 0.18%.

21. Which country had the highest gdp per capita in 2014? What was the gdp per capita?